### ISSUE TITLE
WaterToAirHeatPump:EquationFit sizing companion coil array bounds error

### ISSUE THREAD

**Issue by mjwitte** [NREL/EnergyPlus#9849]
Issue overview
--------------
*This was originally part of #9793 . Splitting into separate issues.*

Another unitary system which attempted to use two separate unitary systems (one for WAHP cooling and one for WAHP heating) fails in debug build with a zero value for companion coil. Fails in `WaterToAirSimple::SizeHVACWaterToAir`, line 2575

```
auto &companionCoolingCoil(state.dataWaterToAirHeatPumpSimple->SimpleWatertoAirHP(simpleWatertoAirHP.CompanionCoolingCoilNum));
```

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 22.2
 - [Unmethours 86434](https://unmethours.com/question/86434/unitary-system-water-air-cc-control-issues/)
 - Helpdesk tickets 16321 and 16322.

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added to EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by mjwitte** [comment]
@rraustad This has been split off from #9793. Draft PR #9814 is still open for reference with some trial heat pump sizing changes. Feel free to take this issue if you want.

**Comment by mjwitte** [comment]
And helpdesk ticket 16739. A possible workaround is to combine the heating/cooling system into one unit, or add dummy heating/cooling coils as needed, so that the system has both coil types.

**Comment by mjwitte** [comment]
See #10735  for a fix for the array bounds error. But the system still doesn't meet setpoint the way it did back in v9.5

For the 16738 file which has some single-sided heat pump system (cooling-only and heating-only), here's the hours not met from v9.5 to v24.2 (with the array bounds fix):
version, occupied hrs ~cooling~ *heating* setpoint not met, occ hrs ~heating~ *cooling* setpoint not met
v9.4, 31, 38
v9.5 31, 38
​v9.6, 31, 3587
v22.1, 31, 3587
v22.2, 560, 3551
v23.1 - allocation error, fatal
v23.2 - allocation error, fatal
v24.1, 632, 30
v24.2, 632, 30
@rraustad Any thoughts here?

**Comment by rraustad** [comment]
There is a couple things wrong here. 1) the WSHP coil sizing expects a pair of coils, it's a HP. So I fixed that. 2) That call to the heating coil where that fix was made sizes the heating coil to 468 W where the cooling coil is much larger at 138309 W. This disparity aside, because I guess you should be able to do that with 2 independent systems, did this defect file run in release to allow you to test in past and present versions of E+? I assume the companionCoil array bounds issue did not show up in release and wonder what was happening in that part of the code because there is an else at the end. What sizes did you get with your single code change? I didn't get to your other question on unmet hours because of this issue.

    Component Sizing Information, AirLoopHVAC:UnitarySystem, UNITARY - CLG, Design Size Cooling Supply Air Flow Rate [m3/s], 1.77558
    Component Sizing Information, AirLoopHVAC:UnitarySystem, UNITARY - CLG, Design Size No Load Supply Air Flow Rate [m3/s], 1.77558
    Component Sizing Information, COIL:COOLING:WATERTOAIRHEATPUMP:EQUATIONFIT, WTR-TO-AIR HP CLG COIL 1, Design Size Rated Air Flow Rate [m3/s], 1.77558
    Component Sizing Information, COIL:COOLING:WATERTOAIRHEATPUMP:EQUATIONFIT, WTR-TO-AIR HP CLG COIL 1, Design Size Rated Total Cooling Capacity [W], 62751.74636
    Component Sizing Information, COIL:COOLING:WATERTOAIRHEATPUMP:EQUATIONFIT, WTR-TO-AIR HP CLG COIL 1, Design Size Rated Sensible Cooling Capacity [W], 77655.40490
    Component Sizing Information, COIL:COOLING:WATERTOAIRHEATPUMP:EQUATIONFIT, WTR-TO-AIR HP CLG COIL 1, Design Size Rated Water Flow Rate [m3/s], 1.09692E-003
    Component Sizing Information, AirLoopHVAC:UnitarySystem, UNITARY - CLG, Design Size Nominal Cooling Capacity [W], 62751.74636
    Component Sizing Information, AirLoopHVAC:UnitarySystem, UNITARY - HTG, Design Size Heating Supply Air Flow Rate [m3/s], 0.55724
    Component Sizing Information, AirLoopHVAC:UnitarySystem, UNITARY - HTG, Design Size No Load Supply Air Flow Rate [m3/s], 0.55724
    Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, WTR-TO-AIR HP HTG COIL 2, Design Size Rated Air Flow Rate [m3/s], 1.77558
    Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, WTR-TO-AIR HP HTG COIL 2, Design Size Rated Heating Capacity [W], 468.82194
    Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, WTR-TO-AIR HP HTG COIL 2, Design Size Rated Water Flow Rate [m3/s], 3.11576E-005
    Component Sizing Information, AirLoopHVAC:UnitarySystem, UNITARY - HTG, Design Size Nominal Heating Capacity [W], 468.82194
    Component Sizing Information, AirLoopHVAC:UnitarySystem, UNITARY - HTG, Design Size Maximum Supply Air Temperature [C], 21.11111

[9849-Solution2-Separate UnitarySystems-in-v24.2-CompanionCoilCrash.txt](https://github.com/user-attachments/files/17059735/9849-Solution2-Separate.UnitarySystems-in-v24.2-CompanionCoilCrash.txt)

**Comment by rraustad** [comment]
I also don't think the call to cooling coil capacity should be there in UnitarySystem because the heating coil was already sized on the call to SimWatertoAirHPSimple at line 3158. This area of code is trying to size the UnitarySystem heating capacity, not the coils. If the cooling coil was present the UnitarySystem would report the size of the cooling coil as the UnitarySystem heating capacity (see down at line 3196 in UnitarySystem). I think the fix is to remove that chunk of code you protected and make sure the WSHP sizing picks up the correct size based on what UnitarySystem told the coil to size to. Maybe the heating coil does that now and if so that code isn't needed, but if that coil model does not do that now it would need to be corrected in the heating coil model. I didn't see that while single stepping but maybe I missed it.

**Comment by rraustad** [comment]
There are some things in sizing that could be looked at.

At the end of the cooling coil capacity sizing is this equation which appears to adjust the capacity with some ratio of coil air flow to system air flow. But this coil is in the "system" so what is this ratio actually doing? It changes the cooling coil size from 138 kW to 60 kW. The dHratio uses the coil mixed air and zone conditions to calculate this ratio. The coil inlet condition should not be adjusted by some ratio of entering condition to the zone condition?

    // adjust for system air flow -- capacity is based on cooling design day calcs
    // adjust by ratio of system to cooling air flow rate and enthalpy delta across the coil at these different airflow
    RatedCapCoolTotalDes *= (RatedAirVolFlowRateDes / CoolingAirVolFlowRateDes) * dHratio;

The coil capacities and the system sizing data line up for cooling but not for heating.

    System Sizing Information, AHU-1, Cooling, Total, 63244.94, 1.77558, 1.77558, NEW.YORK-KENNEDY.INTL.AP ANN CLG 1% CONDNS DB=>MWB, 7/21 15:20:00
    System Sizing Information, AHU-1, Heating, Sensible, 14661.83, 0.43172, 0.43172, NEW.YORK-KENNEDY.INTL.AP ANN HTG 99% CONDNS DB, 1/21 24:00:00

**Comment by mjwitte** [comment]
> I assume the companionCoil array bounds issue did not show up in release and wonder what was happening in that part of the code because there is an else at the end. What sizes did you get with your single code change?

The results listed [above](https://github.com/NREL/EnergyPlus/issues/9849#issuecomment-2347923094) were all with release builds, except the v24.2 result which was debug with my initial fix in place. v23.1 and 23.2 failed with an allocation error even in release. (Note that I've edited that, I had the heating and cooling hrs not met backwards).

I was running a different defect file, apparently, but now I see where the problem is for heating, anyway with that file. The water flow sized to zero.
```
 Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, Design Size Rated Air Flow Rate [m3/s], 0.57633
 Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, User-Specified Rated Air Flow Rate [m3/s], 0.57633
 Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, User-Specified Rated Heating Capacity [W], 10808.49038
 Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, Design Size Rated Water Flow Rate [m3/s], 0.00000
```

**Comment by mjwitte** [comment]
@rraustad Using this branch with your current fix gives the same sizing results. Water flow is still zero, which would explain why it's not heating.

**Comment by rraustad** [comment]
OK, good. The water flow sizing can be corrected now that it is identified.

**Comment by mjwitte** [comment]
FYI, the coil that sizes water flow to zero is hard-sized in every field, except the water flow.
```
  Coil:Heating:WaterToAirHeatPump:EquationFit,
    Perim_1_WSHP_CoilHtg,    !- Name
    OfcSml CWSystem Demand Splitter Outlet Node 1,  !- Water Inlet Node Name
    Perim_1_WSHP_CoilHtg Water Outlet Node,  !- Water Outlet Node Name
    Perim_1_WSHP_CoilClg Unitary Outlet Node,  !- Air Inlet Node Name
    Perim_1_WSHP_CoilHtg Unitary Outlet Node,  !- Air Outlet Node Name
    0.576328059212544,       !- Rated Air Flow Rate {m3/s}
    Autosize,                !- Rated Water Flow Rate {m3/s}
    10808.4903750586,        !- Gross Rated Heating Capacity {W}
    4.93922286267479,        !- Gross Rated Heating COP {W/W}
```

**Comment by rraustad** [comment]
We are using different test files, point me to the defect file you are using.

**Comment by mjwitte** [comment]
See https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/9000s/9849/Ticket%2016738

**Comment by rraustad** [comment]
No idea why someone would set the water flow rate to 0 if the capacity was not autosized. This is now fixed in #10735.

    Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, Design Size Rated Air Flow Rate [m3/s], 0.57633
    Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, User-Specified Rated Air Flow Rate [m3/s], 0.57633
    Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, User-Specified Rated Heating Capacity [W], 10808.49038
    Component Sizing Information, COIL:HEATING:WATERTOAIRHEATPUMP:EQUATIONFIT, PERIM_1_WSHP_COILHTG, Design Size Rated Water Flow Rate [m3/s], 4.67258E-004
